### ISSUE TITLE
New i_q current model with modeled friction

### ISSUE THREAD

**Issue by AndPatr** [ADVRHumanoids/awesome_leg#5]
We have implemented a new and very simple model for the quadrature component of the current which also accounts for static and dynamic friction.
The i_q is described by the following very simple model:
 ![](https://user-images.githubusercontent.com/90312267/151814847-38f18bb1-0dc2-49af-9f05-ccb2a3e100bd.jpg).
The model was fitted over a calibration trajectory and the resulting friction coefficients were also validated on different trajectories, showing promising results.

The obtained coefficients are:
- Hip joint: Kd0 = 2.2 N , Kd1 = 2.4 N/(rad/s)
- Knee joint: Kd0 = 6.8 N, Kd1 = 4.9 N/(rad/s)

These values are coherent with the higher friction showed by the knee joint.

Results on the calibration trajectory:
 - Joint positions (ref VS measured): ![q_p](https://user-images.githubusercontent.com/90312267/151816008-bddb9262-f85c-402c-8773-ac50aa751ecc.png)
 - Joint velocities (ref VS measured): ![q_p_dot](https://user-images.githubusercontent.com/90312267/151816143-f86f8f8c-3c1f-49ad-94b2-3038f561f31d.png)
 - Measured VS estimated i_q: ![i_q](https://user-images.githubusercontent.com/90312267/151816204-4c4e6ee0-387d-45f9-a78f-01e9914da3ea.png)
Results on test trajectory 1:
- Joint positions (ref VS measured): ![q_p](https://user-images.githubusercontent.com/90312267/151816382-2a42d6cc-e51d-44e8-9867-dcc4b9d30dd3.png)
- Joint velocities (ref VS measured): ![q_p_dot](https://user-images.githubusercontent.com/90312267/151816452-f21713f2-9700-4588-be0f-da1613f94d51.png)
- Measured VS estimated i_q: ![i_q](https://user-images.githubusercontent.com/90312267/151816512-f32b8718-9709-4ea9-aa67-d728022fefeb.png)

Results on test trajectory 2:
- Joint positions (ref VS measured): ![q_p](https://user-images.githubusercontent.com/90312267/151817849-9b501b9c-8e79-476d-8376-adc20d937a07.png)
- Joint velocities (ref VS measured): ![q_p_dot](https://user-images.githubusercontent.com/90312267/151817917-4c5835f1-261c-4a0c-be56-987990b91350.png)
- Measured VS estimated i_q:  ![i_q](https://user-images.githubusercontent.com/90312267/151817950-b25748bf-8400-4d33-aba5-13a99feac993.png)

**Comment by ntsagarakis** [comment]
Very nice. is this fed as feedforward current to the motor driver? 
One comment on the model. The terns of second part inside the bracket are also using link side variables.  Why the gear ratio "\mu" is in the numerator and not in the denominator as in the inertia first term to scale down this to the motor side ?

**Comment by AndPatr** [comment]
This is just a model for estimating the i_q current, given a joint trajectory and the torques on the link side, so it was not used as a feedforward.  It will be integrated into the optimization process to account for the increase in current brought by friction.

The plots you see employ joint accelerations (obtained by differentiation and filtering), velocities and measured link torques (which were also validated) to compute the left-hand side of the relationship. As you see, we have a relatively good correspondence between the estimated and the measured i_q currents. This gives us confidence that the obtained friction parameters are indeed good and we can use them into the optimization (or also somewhere else).

If I understood correctly, probably your question stems from assuming a different convention from the one I used here.
My \mu is <1 (so e.g. 1/50). This means that the link side torques are scaled _down_ to the motor side, while the q_p_ddot (link-side variable) is scaled _up_ . Does this answer your question?

Ps.
I chose to model the friction torques as if they were acting on the link side (so they are scaled down by \mu), but this is actually arbitrary and inconsequential (the \mu/\eta term could be simply integrated into the friction coefficients).

**Comment by ntsagarakis** [comment]
ok all clear. thanks ! Can you briefly comment on the  identification of the friction coefficients? what experiments you executed for their estimation? Also what the efficiency coefficient?

**Comment by AndPatr** [comment]
Yes, sure. 
We have a series of trajectories (basically simil-trots and swing maneuvers) tested on the leg during last week and also the associated logged data (measurements, rosbags, etc...). The identification of the parameters was carried out using a single trajectory, which was selected among the others available. In particular, if you look at the q_p plots, this is a swing maneuver of the leg (a faster version of this one). This trajectory is particularly suitable for calibration, since it has both static and dynamic phases. 

The computation of the friction coefficients was made setting up a regression over the i_q measured during the test. If you see the formula, it is linear in the required coefficients, so a simple QP was used (actually, the QP also employs the J_r and K_t as optimization variables, but these were fixed to their nominal values in this case).

As for the efficiency, here I set it to 1 for both axis and I let the Kd0 and Kd1 account for all the friction; in my opinion, this is the choice which makes most sense from a physical point of view.
